## Summary:  
Software: openstack/nova  
## Fix:  
Commit: https://github.com/openstack/nova/commit/8b289237ed6d53738c22878decf0c429301cf3d0  
Change Size: +37 -4  
Author: Matt Riedemann (IBM)  
Fix Time (of the Last Commit): 2015-11-18  
  
Fix Pattern: Sanitization -> Redaction -> Reuse -> Keyword/RegexBased -> RedactionUtilDefined  
Fixing-commits Count: 1  
Total Change Size: +37 -4  
## Sample logs: No  
## Logging statements:  
### Source labels:  
Type of Sensitive Information:authentication data -> sensitive password information  
### Sink labels:  
#### Sink 1:  
Logging call Function:  Third-party Libaries（OpenStack oslo.log）  
Sensitiveness-Identifiable Source:  Same Modules & Cross Files（https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/volume_utils.py#L96） -> String Literals（auth_password，finally outputed to logs）  
#### Sink 2:  
Logging call Function:  Third-party Libaries（OpenStack oslo.log）  
Sensitiveness-Identifiable Source:  Same Modules & Cross Files（https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/volume_utils.py#L96） -> String Literals（auth_password，finally outputed to logs）  
Programming Language:Python  
```Python  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/volumeops.py#L93-L109  
  
    def _connect_hypervisor_to_volume(self, sr_ref, connection_data):  
        LOG.debug("Connect volume to hypervisor: %s", connection_data) # HERE IS THE SINK 1  
        if 'vdi_uuid' in connection_data:  
            vdi_ref = volume_utils.introduce_vdi(  
                    self._session, sr_ref,  
                    vdi_uuid=connection_data['vdi_uuid'])  
  
        elif 'target_lun' in connection_data:  
            vdi_ref = volume_utils.introduce_vdi(  
                    self._session, sr_ref,  
                    target_lun=connection_data['target_lun'])  
  
        else:  
            # NOTE(sirp): This will introduce the first VDI in the SR  
            vdi_ref = volume_utils.introduce_vdi(self._session, sr_ref)  
  
        return vdi_ref  
  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/volume_utils.py#L62-L98  
  
def _parse_volume_info(connection_data):  
    """Parse device_path and mountpoint as they can be used by XenAPI.  
    In particular, the mountpoint (e.g. /dev/sdc) must be translated  
    into a numeric literal.  
    """  
    volume_id = connection_data['volume_id']  
    target_portal = connection_data['target_portal']  
    target_host = _get_target_host(target_portal)  
    target_port = _get_target_port(target_portal)  
    target_iqn = connection_data['target_iqn']  
  
    log_params = {  
        "vol_id": volume_id,  
        "host": target_host,  
        "port": target_port,  
        "iqn": target_iqn  
    }  
    LOG.debug('(vol_id,host,port,iqn): '  
              '(%(vol_id)s,%(host)s,%(port)s,%(iqn)s)', log_params)  
  
    if (volume_id is None or  
        target_host is None or  
            target_iqn is None):  
        raise exception.StorageError(  
                reason=_('Unable to obtain target information %s') %  
                        connection_data) # HERE IS THE INNER SINK  
    volume_info = {}  
    volume_info['id'] = volume_id  
    volume_info['target'] = target_host  
    volume_info['port'] = target_port  
    volume_info['targetIQN'] = target_iqn  
    if ('auth_method' in connection_data and  
            connection_data['auth_method'] == 'CHAP'):  
        volume_info['chapuser'] = connection_data['auth_username']  
        volume_info['chappassword'] = connection_data['auth_password']  
  
    return volume_info  
  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/volume_utils.py#L44-L59  
def parse_sr_info(connection_data, description=''):  
    label = connection_data.pop('name_label',  
                                'tempSR-%s' % connection_data.get('volume_id'))  
    params = {}  
    if 'sr_uuid' not in connection_data:  
        params = _parse_volume_info(connection_data) # HERE IS THE FLOW  
        # This magic label sounds a lot like 'False Disc' in leet-speak  
        uuid = "FA15E-D15C-" + str(params['id'])  
    else:  
        uuid = connection_data['sr_uuid']  
        for k in connection_data.get('introduce_sr_keys', {}):  
            params[k] = connection_data[k]  
    params['name_description'] = connection_data.get('name_description',  
                                                     description)  
  
    return (uuid, label, params)  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/volumeops.py#L83-L91  
    def _connect_to_volume_provider(self, connection_data, instance_name):  
        sr_uuid, sr_label, sr_params = volume_utils.parse_sr_info(  
                connection_data, 'Disk-for:%s' % instance_name) # HERE IS THE FLOW  
        sr_ref = volume_utils.find_sr_by_uuid(self._session, sr_uuid)  
        if not sr_ref:  
            # introduce SR because not already present  
            sr_ref = volume_utils.introduce_sr(  
                    self._session, sr_uuid, sr_label, sr_params)  
        return (sr_ref, sr_uuid)  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/volumeops.py#L52-L76  
    def _attach_volume(self, connection_info, vm_ref=None, instance_name=None,  
                       dev_number=None, hotplug=False):  
  
        self._check_is_supported_driver_type(connection_info)  
  
        connection_data = connection_info['data']  
        sr_ref, sr_uuid = self._connect_to_volume_provider(connection_data,  
                                                           instance_name) # HERE IS THE FLOW  
        try:  
            vdi_ref = self._connect_hypervisor_to_volume(sr_ref,  
                                                         connection_data)  
            vdi_uuid = self._session.VDI.get_uuid(vdi_ref)  
            LOG.info(_LI('Connected volume (vdi_uuid): %s'), vdi_uuid)  
  
            if vm_ref:  
                self._attach_volume_to_vm(vdi_ref, vm_ref, instance_name,  
                                          dev_number, hotplug)  
  
            return (sr_uuid, vdi_uuid)  
        except Exception:  
            with excutils.save_and_reraise_exception():  
                # NOTE(sirp): Forgetting the SR will have the effect of  
                # cleaning up the VDI and VBD records, so no need to handle  
                # that explicitly.  
                volume_utils.forget_sr(self._session, sr_ref)  
  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/volumeops.py#L38-L46  
  
    def attach_volume(self, connection_info, instance_name, mountpoint,  
                      hotplug=True):  
        """Attach volume to VM instance."""  
        # TODO(johngarbutt) move this into _attach_volume_to_vm  
        dev_number = volume_utils.get_device_number(mountpoint)  
  
        vm_ref = vm_utils.vm_ref_or_raise(self._session, instance_name)  
        return self._attach_volume(connection_info, vm_ref,  
                                   instance_name, dev_number, hotplug) # HERE IS THE FLOW  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/vmops.py#L215-L229  
    def _attach_mapped_block_devices(self, instance, block_device_info):  
        # We are attaching these volumes before start (no hotplugging)  
        # because some guests (windows) don't load PV drivers quickly  
        block_device_mapping = virt_driver.block_device_info_get_mapping(  
                block_device_info)  
        for vol in block_device_mapping:  
            if vol['mount_device'] == instance['root_device_name']:  
                # NOTE(alaski): The root device should be attached already  
                continue  
            connection_info = vol['connection_info']  
            mount_device = vol['mount_device'].rpartition("/")[2]  
            self._volumeops.attach_volume(connection_info,  
                                          instance['name'],  
                                          mount_device,  
                                          hotplug=False) # HERE IS THE FLOW  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/vmops.py#L237-L263  
    def _restore_orig_vm_and_cleanup_orphan(self, instance,  
                                            block_device_info=None,  
                                            power_on=True):  
        # NOTE(sirp): the original vm was suffixed with '-orig'; find it using  
        # the old suffix, remove the suffix, then power it back on.  
        name_label = self._get_orig_vm_name_label(instance)  
        vm_ref = vm_utils.lookup(self._session, name_label)  
  
        # NOTE(danms): if we're reverting migration in the failure case,  
        # make sure we don't have a conflicting vm still running here,  
        # as might be the case in a failed migrate-to-same-host situation  
        new_ref = vm_utils.lookup(self._session, instance['name'])  
        if vm_ref is not None:  
            if new_ref is not None:  
                self._destroy(instance, new_ref)  
            # Remove the '-orig' suffix (which was added in case the  
            # resized VM ends up on the source host, common during  
            # testing)  
            name_label = instance['name']  
            vm_utils.set_vm_name_label(self._session, vm_ref, name_label)  
            self._attach_mapped_block_devices(instance, block_device_info) # HERE IS THE FLOW  
        elif new_ref is not None:  
            # We crashed before the -orig backup was made  
            vm_ref = new_ref  
  
        if power_on and vm_utils.is_vm_shutdown(self._session, vm_ref):  
            self._start(instance, vm_ref)  
  
# https://github.com/openstack/nova/blob/f0f91cb63ba7cefdbc17a609362efc9f2fb61a5e/nova/virt/xenapi/vmops.py#L1155-L1161  
            try:  
                self._restore_orig_vm_and_cleanup_orphan(instance)  
                # TODO(johngarbutt) should also cleanup VHDs at destination  
            except Exception as rollback_error:  
                LOG.warning(_LW("_migrate_disk_resizing_up failed to "  
                                "rollback: %s"), rollback_error,  
                            instance=instance) # HERE IS THE SINK 2  
  
```  
